monkey-patches.md

```diff
@@ -0,0 +1,144 @@
+
+For example, in order to ensure you have the original window, jquery, and undefined variables you can
+
+```javascript
+function(window, $, undefined) {
+  // your code
+}(window, $)
+```
+Now your code will continue working even if the user will do something like
+```javascript
+window.$ = { ajax: function() {} };
+```
+
+This method provides a very partial protection. Lets say that after your code is loaded the user adds the following code:
+```javascript
+  $.ajax = function() {}
+```
+
+Since you only kept a *reference* to the jquery variable, you are still vulnerable to modification on the object itself.
+
+In reality this method has other benefits, such as shortening the list of scopes JS needs to look in to find your variable, or enabling js obfuscators to rename common variables such as window or document.
+
+### More IIFE
+
+We can utilize a similar construct to [UMDs](https://github.com/umdjs/umd) module pattern. Here we clone the objects we want to protect, and then pass it to our IIFE:
+
+```javascript
+function(factory) {
+  var $ = clone($);
+
+  factory(window, $)
+}(function(window, $){
+  // your code
+})
+```
+
+Of course we have to ask our self - where is the limit? Should we just clone window and be done with it? Should take the burden of managing a list of all language function we use and clone them by an individual basis?
+
+
+Alas, even this method would not provide real guarantee for native function integrity, as the user might override them before loading our code. It is in fact impossible to create such security
+
+### Detecting a monkey patched method?
+Wouldn't it be nice to just be able to test weather a function have been monkey patched? If we'll be able to do that, we'll at least be able to fail gracefully. JavaScript won't make our life easy in this task as well. A property that might come in handy is the output of the toString method for native functions:
+
+```javascript
+  Function.prototype.bind.toString();
+  // function bind() { [native code] }
+```
+
+Utilizing this we can create a monkey patch detection method:
+```javascript
+  function isMonkeyPatched(fn) {
+    return !Function.prototype.toString.call(fn).match(/native code/);
+  }
+```
+
+Unfortunately this will only work for native functions, and will fail if our beloved users monkey patch toString.
+
+
+In conclusion
+--------------
+Javascript, as wonderful language that it is allows everyone to override almost everything. We have encountered some very bizarre bugs that resulted from strange monkey patches. Even though it's always nice to find out that a bug was not caused by your code, you need to ask yourself what's next? Do you approach the client and tell them their code is not good? Or do you handle it yourself?
```

Comment:
Good point on the 'breaks our code'. I think it should stay 'client' and 'them'. Behind a single client there are several people. Anyway it's not politically correct to write 'him' because the client might be a woman. I didn't invert this I swear...